### PR TITLE
HPCC-29480 Check foreign scope in DFU server when foreigncopy

### PR DESCRIPTION
**Pull request by wangkx**
## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: dali/dfu/dfurun.cpp [hunk 1]
```diff
@@ -1526,7 +1559,7 @@ class CDFUengine: public CInterface, implements IDFUengine
                         if (needrep)
                             feedback.repmode=cProgressReporter::REPbefore;
                         if (foreigncopy)
-                            checkPhysicalFilePermissions(srcFdesc,userdesc,false);
+                            checkForeignFilePermissions(source,userdesc,false);
```
Comment (jakesmith) [inline dali/dfu/dfurun.cpp:1562]: looks fine, but unfortunately, we need to keep backward compatibility, by also falling back to perform the legacy check on a the physical file format (i.e. continue to call checkPhysicalFilePermissions if denied on the new semantics.
But it should be a config. option so we can disable these legacy checks (default to off/no checking of legacy in cloud).

**Review by jakesmith — COMMENTED** [review]
@wangkx - looks good, but see comment, re. handling legacy via option (dfuserver option)

File: dali/dfu/dfurun.cpp [hunk 2]
```diff
@@ -601,6 +601,44 @@ class CDFUengine: public CInterface, implements IDFUengine
...
+        else
+            foreignuserdesc.set(user);
+
+        SecAccessFlags perm = queryDistributedFileDirectory().getDLFNPermissions(dlfn,foreignuserdesc,auditflags);
+
+        bool checkLegacyPhysicalPerms = getGlobalConfigSP()->getPropBool("expert/@failOverToLegacyPhysicalPerms", !isContainerized());
+        if ((!HASREADPERMISSION(perm) || (write&&!HASWRITEPERMISSION(perm))) && checkLegacyPhysicalPerms)
+            perm = queryDistributedFileDirectory().getFDescPermissions(fd,user,auditflags);
+
+        if (!HASREADPERMISSION(perm))
```
Comment (jakesmith) [inline dali/dfu/dfurun.cpp:636]: does a user necessarily need read permissions to write a file?
As it's coded both are needed if write=true, and the error you get (if no read perms) (on a write=true) is "Lookup permission denied .." - that may be confusing.

**Review by jakesmith — COMMENTED** [review]
@wangkx - 1 question.

File: dali/dfu/dfurun.cpp (hunk 2 above)
Comment (wangkx) [reply dali/dfu/dfurun.cpp:636]: This is copied from the checkPhysicalFilePermissions(). Looks like it assumes that read permission is needed to write a file. I will check !write here.

**Comment by wangkx** [comment]
@jakesmith please review

File: dali/dfu/dfurun.cpp [hunk 3]
```diff
@@ -595,12 +595,50 @@ class CDFUengine: public CInterface, implements IDFUengine
             auditflags |= DALI_LDAP_WRITE_WANTED;
         SecAccessFlags perm = queryDistributedFileDirectory().getFDescPermissions(fd,user,auditflags);
         IDFS_Exception *e = NULL;
-        if (!HASREADPERMISSION(perm))
+        if (!write&&!HASREADPERMISSION(perm))
             throw MakeStringException(DFSERR_LookupAccessDenied,"Lookup permission denied for physical file(s)");
```
Comment (jakesmith) [inline dali/dfu/dfurun.cpp:599]: trivial: but worth changing these to makeStringExceptionV while here (include filename)

File: dali/dfu/dfurun.cpp [hunk 4]
```diff
@@ -595,12 +595,50 @@ class CDFUengine: public CInterface, implements IDFUengine
             auditflags |= DALI_LDAP_WRITE_WANTED;
         SecAccessFlags perm = queryDistributedFileDirectory().getFDescPermissions(fd,user,auditflags);
         IDFS_Exception *e = NULL;
-        if (!HASREADPERMISSION(perm))
+        if (!write&&!HASREADPERMISSION(perm))
```
Comment (jakesmith) [inline dali/dfu/dfurun.cpp:598]: it's fine, but simpler if:
```
        if (write)
        {
            if (!HASWRITEPERMISSION(perm))
                throw MakeStringException(DFSERR_CreateAccessDenied,"Create permission denied for physical file(s)");
       }
       else if (!HASREADPERMISSION(perm))
            throw MakeStringException(DFSERR_LookupAccessDenied,"Lookup permission denied for physical file(s)");
 ```

File: dali/dfu/dfurun.cpp [hunk 5]
```diff
@@ -595,12 +595,50 @@ class CDFUengine: public CInterface, implements IDFUengine
...
+        else
+            foreignuserdesc.set(user);
+
+        SecAccessFlags perm = queryDistributedFileDirectory().getDLFNPermissions(dlfn,foreignuserdesc,auditflags);
+
+        bool checkLegacyPhysicalPerms = getGlobalConfigSP()->getPropBool("expert/@failOverToLegacyPhysicalPerms", !isContainerized());
+        if (((!write&&!HASREADPERMISSION(perm)) || (write&&!HASWRITEPERMISSION(perm))) && checkLegacyPhysicalPerms)
+            perm = queryDistributedFileDirectory().getFDescPermissions(fd,user,auditflags);
+
+        if (!write&&!HASREADPERMISSION(perm))
```
Comment (jakesmith) [inline dali/dfu/dfurun.cpp:636]: see prev. comment re. if (write) 1st.

File: dali/dfu/dfurun.cpp [hunk 6]
```diff
@@ -595,12 +595,50 @@ class CDFUengine: public CInterface, implements IDFUengine
...
+            throw makeStringException(DFSERR_LookupAccessDenied,"Lookup permission denied for physical file(s)");
```
[the remark this answers is not in the document]
Comment (wangkx) [reply dali/dfu/dfurun.cpp:637]: At the beginning of the checkForeignFilePermissions(), a filename is read by fSpec->getLogicalName(). Should I use that filename here?

File: dali/dfu/dfurun.cpp (hunk 3 above)
Comment (wangkx) [reply dali/dfu/dfurun.cpp:599]: What if there are multiple files in the fd? Should I use fd->getTraceName() to get a name (filename?) here?

Comment (wangkx) [reply dali/dfu/dfurun.cpp:599]: I guess I can use the name returned by fd->getTraceName() which returns all file names if multiple.

Comment (jakesmith) [reply dali/dfu/dfurun.cpp:599]: might be big..
maybe spot the multi case, and truncate if >255 with "..."

File: dali/dfu/dfurun.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (jakesmith) [reply dali/dfu/dfurun.cpp]: yes that's fine.

**Review by jakesmith — DISMISSED** [review]
@wangkx - looks ok, please squash.

**Comment by wangkx** [comment]
@jakesmith I did squash.

File: dali/dfu/dfurun.cpp [hunk 7]
```diff
@@ -593,12 +593,63 @@ class CDFUengine: public CInterface, implements IDFUengine
...
         if (write)
             auditflags |= DALI_LDAP_WRITE_WANTED;
+
+        StringBuffer traceName;
+        fd->getTraceName(traceName);
+        if (traceName.length() > 255)
+        {
+            traceName.setLength(255);
+            traceName.append("...");
+        }
```
Comment (jakesmith) [inline dali/dfu/dfurun.cpp:603]: This block is not needed unless access is denied, should be moved inside conditional.

**Review by jakesmith — COMMENTED** [review]
@wangkx - noticed 1 more thing.

**Comment by wangkx** [comment]
@jakesmith please review

**Review by jakesmith — COMMENTED** [review]
There are overlapping comments in another PR (https://github.com/hpcc-systems/HPCC-Platform/pull/17528), that relate to this one.

Shall we get this one re-approved 1st?
Otherwise there's duplication of [reviewing] effort.

**Comment by wangkx** [comment]
@jakesmith The previous changes are rolled back. I also move the code. Please review.

**Review by jakesmith — COMMENTED** [review]
@wangkx - looks good. Please squash

**Review by jakesmith — APPROVED** [review]

**Pull request merged**